src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WinHttpWebSocket.cs

```diff
@@ -565,6 +577,134 @@ private void UpdateServerCloseStatus()
             }
         }
 
+        private void VerifyUpgradeResponse()
+        {
+            // Check the status code
+            var statusCode = GetHttpStatusCode();
+            if (statusCode != HttpStatusCode.SwitchingProtocols)
+            {
+                Abort();
+                return;
+            }
+
+            _subProtocol = GetResponseHeaderStringInfo(HeaderNameWebSocketProtocol);
+        }
+
+        private void AddRequestHeaders(Uri uri, ClientWebSocketOptions options)
+        {
+            var requestHeadersBuffer = new StringBuilder();
+
+            // Manually add cookies.
+            if (options.Cookies != null)
+            {
+                string cookieHeader = GetCookieHeader(uri, options.Cookies);
+                if (!string.IsNullOrEmpty(cookieHeader))
+                {
+                    requestHeadersBuffer.AppendLine(cookieHeader);
+                }
+            }
+
+            // Serialize general request headers.
+            requestHeadersBuffer.AppendLine(options.RequestHeaders.ToString());
+
+            var subProtocols = options.RequestedSubProtocols;
+            if (subProtocols.Count > 0)
+            {
+                requestHeadersBuffer.AppendLine(string.Format("{0}: {1}", HeaderNameWebSocketProtocol,
+                    string.Join(", ", subProtocols)));
```

Comment:
From a perf perspective, it'd be better if these intermediate strings were avoided, e.g. by appending HeaderNameWebSocketProtocol, then appending a colon, then appending each of the subProtocols individually, etc.